File: app-java/src/main/java/com/tritiumstudios/phasmophobiaevidencetool/android/data/viewmodels/shared/GlobalPreferencesViewModel.java
package com.tritiumstudios.phasmophobiaevidencetool.android.data.viewmodels.shared;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.tritiumstudios.phasmophobiaevidencetool.R;
import com.tritiumstudios.phasmophobiaevidencetool.android.data.controllers.ReviewTrackingData;
import com.tritiumstudios.phasmophobiaevidencetool.android.data.controllers.theming.CustomTheme;
import com.tritiumstudios.phasmophobiaevidencetool.android.data.controllers.theming.subsets.ColorThemeControl;
import com.tritiumstudios.phasmophobiaevidencetool.android.data.controllers.theming.subsets.FontThemeControl;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Locale;

/** @noinspection SameParameterValue*/
public class GlobalPreferencesViewModel extends SharedViewModel {

    // Review Tracker
    private ReviewTrackingData reviewRequestData;

    // Language
    //private LanguageControl languageControl;
    private String languageName = Locale.getDefault().getLanguage();

    // Persistent Styles
    private FontThemeControl fontThemeControl;
    private ColorThemeControl colorThemeControl;

    // Generic settings
    private int huntWarningFlashTimeout = -1;
    private boolean
            isAlwaysOn = false,
            isHuntAudioAllowed = true,
            networkPreference = true,
            isLeftHandSupportEnabled = false;

    // Title screen increments
    private boolean canShowIntroduction = true;

    @Override
    public void setFileName() {
        fileName = R.string.preferences_globalFile_name;
    }

    /**
     * init method
     *
     * @param context
     */
    public boolean init(@NonNull Context context) {

        setFileName();

        SharedPreferences sharedPref = getSharedPreferences(context);

        setNetworkPreference(sharedPref.getBoolean(context.getResources().getString(R.string.preference_network), getNetworkPreference()));
        setLanguageName(sharedPref.getString(context.getResources().getString(R.string.preference_language), getLanguageName()));
        setIsAlwaysOn(sharedPref.getBoolean(context.getResources().getString(R.string.preference_isAlwaysOn), getIsAlwaysOn()));
        setHuntWarningAudioAllowed(sharedPref.getBoolean(context.getResources().getString(R.string.preference_isHuntAudioWarningAllowed), getIsHuntAudioAllowed()));
        setHuntWarningFlashTimeout(sharedPref.getInt(context.getResources().getString(R.string.preference_huntWarningFlashTimeout), getHuntWarningFlashTimeout()));

        setLeftHandSupportEnabled(sharedPref.getBoolean(context.getResources().getString(R.string.preference_isLeftHandSupportEnabled), getIsLeftHandSupportEnabled()));
        setCanShowIntroduction(sharedPref.getBoolean(context.getResources().getString(R.string.tutorialTracking_canShowIntroduction), getCanShowIntroduction()));

        reviewRequestData = new ReviewTrackingData(
                sharedPref.getBoolean(context.getResources().getString(R.string.reviewtracking_canRequestReview), false),
                sharedPref.getLong(context.getResources().getString(R.string.reviewtracking_appTimeAlive), 0),
                sharedPref.getInt(context.getResources().getString(R.string.reviewtracking_appTimesOpened), 0)
        );

        fontThemeControl = new FontThemeControl(context);
        fontThemeControl.init(
                sharedPref.getString(
                        context.getResources().getString(R.string.preference_savedFont),
                        getFontThemeID()
                )
        );

        colorThemeControl = new ColorThemeControl(context);
        colorThemeControl.init(
                sharedPref.getString(
                        context.getResources().getString(R.string.preference_savedTheme),
                        getColorThemeID()
            )
        );

        saveToFile(context);

        return true;
    }

    public void setLeftHandSupportEnabled(boolean isLeftHandSupportEnabled) {
        this.isLeftHandSupportEnabled = isLeftHandSupportEnabled;
    }

    public boolean getIsLeftHandSupportEnabled() {
        return isLeftHandSupportEnabled;
    }

    public void setNetworkPreference(boolean preference) {
        this.networkPreference = preference;
    }

    public boolean getNetworkPreference() {
        return networkPreference;
    }

    public void incrementAppOpenCount(@NonNull Context context) {

        getReviewRequestData().incrementTimesOpened();

        try {
            saveTimesOpened(context, getEditor(context), true);
        } catch (NullPointerException e) {
            e.printStackTrace();
        }

    }

    /**
     * getReviewRequestData method
     *
     * @return reviewRequestData
     */
    public ReviewTrackingData getReviewRequestData() {
        return reviewRequestData;
    }


    /**
     * setLanguageName method
     *
     * @param languageName
     */
    public void setLanguageName(String languageName) {
        this.languageName = languageName;
    }

    /**
     * getLanguageName method
     *
     * @return languageName
     */
    public String getLanguageName() {
        return languageName;
    }

    /**
     * getLanguage
     * <p>
     * Gets the language saved to GlobalPreferences.
     * Defaults to return 'en' if there is no previously saved preference.
     *
     * @return The language specified in the Preferences data, or otherwise English
     */
    @NonNull
    public String getLanguage(@NonNull Context context) {
        String lang = context.getSharedPreferences(
                context.getResources().getString(fileName), Context.MODE_PRIVATE).getString(
                "chosenLanguage", "en");

        Log.d("Current Chosen Language", lang);

        return lang;
    }

    public int getLanguageIndex(@NonNull ArrayList<String> languageNames) {
        for(int i = 0; i < languageNames.size(); i++) {
            if(getLanguageName().equalsIgnoreCase(languageNames.get(i))) {
                return i;
            }
        }
        return 0;
    }

    /**
     * setLanguage method
     *
     * @param position
     * @param languageNames
     */
    public void setLanguage(int position, @NonNull String[] languageNames) {
        if(position < 0 || position >= languageNames.length) { return; }

        setLanguageName(languageNames[position]);
    }


    /**
     * getIsAlwaysOn method
     *
     * @return isAlwaysOn
     */
    public boolean getIsAlwaysOn() {
        return isAlwaysOn;
    }

    /**
     * setIsAlwaysOn method
     *
     * @param isAlwaysOn
     */
    public void setIsAlwaysOn(boolean isAlwaysOn) {
        this.isAlwaysOn = isAlwaysOn;
    }

    /**
     * setHuntWarningAudioAllowed method
     *
     * @param isAllowed
     */
    public void setHuntWarningAudioAllowed(boolean isAllowed) {
        isHuntAudioAllowed = isAllowed;
    }

    /**
     * getIsHuntAudioAllowed method
     *
     * @return isHuntAudioAllowed
     */
    public boolean getIsHuntAudioAllowed() {
        return isHuntAudioAllowed;
    }

    /**
     * @return
     */
    public boolean isHuntWarningAudioAllowed() {
        return isHuntAudioAllowed;
    }

    /**
     * setHuntWarningFlashTimeout method
     *
     * @param timeout
     */
    public void setHuntWarningFlashTimeout(int timeout) {
        huntWarningFlashTimeout = timeout;
    }

    /**
     * getHuntWarningFlashTimeout method
     *
     * @return huntWarningFlashTimeout
     */
    public int getHuntWarningFlashTimeout() {
        return huntWarningFlashTimeout;
    }

    /**
     * getColorSpace method
     *
     * @return ColorSpace
     */
    /*
    public int getColorTheme() {
        return colorThemeControl.getSavedIndex();
    }
    */
    public String getColorThemeID() {
        return colorThemeControl.getID();
    }
    public CustomTheme getColorTheme() {
        return colorThemeControl.getCurrentTheme();
    }

    /**
     * getFontType method
     *
     * @return fontType
     */
    /*
    public int getFontTheme() {
        return fontThemeControl.getSavedIndex();
    }
    */
    public String getFontThemeID() {
        return fontThemeControl.getID();
    }
    public CustomTheme getFontTheme() {
        return fontThemeControl.getCurrentTheme();
    }

    public FontThemeControl getFontThemeControl() {
        return fontThemeControl;
    }

    public ColorThemeControl getColorThemeControl() {
        return colorThemeControl;
    }

    public void setCanShowIntroduction(boolean canShowIntroduction) {
        this.canShowIntroduction = canShowIntroduction;
    }

    public boolean getCanShowIntroduction() {
        return canShowIntroduction;
    }

    public boolean canShowIntroduction() {
        return canShowIntroduction && reviewRequestData.getTimesOpened() <= 1;
    }

    /**
     *
     * @param c
     * @param editor
     * @param localApply
     */
    private void saveNetworkPreference(
            @NonNull Context c, @Nullable SharedPreferences.Editor editor, boolean localApply) {
        if(editor == null && (editor = getEditor(c)) == null) { return; }

        editor.putBoolean(
                c.getResources().getString(R.string.preference_network),
                getNetworkPreference());
        if(localApply) {
            editor.apply();
        }

    }

    /**
     *
     * @param c
     * @param editor
     * @param localApply
     */
    private void saveChosenLanguage(
            @NonNull Context c, @Nullable SharedPreferences.Editor editor, boolean localApply) {
        if(editor == null && (editor = getEditor(c)) == null) { return; }

        editor.putString(
                c.getResources().getString(R.string.preference_language),
                getLanguageName());

        if (localApply) {
            editor.apply();
        }
    }

    /**
     *
     * @param c
     * @param editor
     * @param localApply
     */
    private void saveAlwaysOnState(
            @NonNull Context c, @Nullable SharedPreferences.Editor editor, boolean localApply) {
        if(editor == null && (editor = getEditor(c)) == null) { return; }

        editor.putBoolean(
                c.getResources().getString(R.string.preference_isAlwaysOn),
                getIsAlwaysOn());

        if(localApply) {
            editor.apply();
        }
    }

    /**
     *
     * @param c
     * @param editor
     * @param localApply
     */
    private void saveHuntWarningAudioAllowed(
            @NonNull Context c, @Nullable SharedPreferences.Editor editor, boolean localApply) {
        if(editor == null && (editor = getEditor(c)) == null) { return; }

        editor.putBoolean(
                c.getResources().getString(R.string.preference_isHuntAudioWarningAllowed),
                getIsHuntAudioAllowed());

        if(localApply) {
            editor.apply();
        }
    }

    /**
     *
     * @param c
     * @param editor
     * @param localApply
     */
    private void saveHuntWarningFlashTimeout(
            @NonNull Context c, @Nullable SharedPreferences.Editor editor, boolean localApply) {
        if(editor == null && (editor = getEditor(c)) == null) { return; }

        editor.putInt(
                c.getResources().getString(R.string.preference_huntWarningFlashTimeout),
                getHuntWarningFlashTimeout());

        if(localApply) {
            editor.apply();
        }
    }

    public void saveColorSpace(@NonNull Context c) {
        saveColorSpace(c, getEditor(c), true);
    }

    /**
     *
     * @param c
     * @param editor
     * @param localApply
     */
    private void saveColorSpace(
            @NonNull Context c, @Nullable SharedPreferences.Editor editor, boolean localApply) {
        if(editor == null && (editor = getEditor(c)) == null) { return; }

        editor.putString(c.getResources().getString(R.string.preference_savedTheme), getColorThemeID());

        if(localApply) {
            editor.apply();
        }
    }

    /**
     *
     * @param c
     * @param editor
     * @param localApply
     */
    private void saveFontType(
            @NonNull Context c, @Nullable SharedPreferences.Editor editor, boolean localApply) {
        if(editor == null && (editor = getEditor(c)) == null) { return; }

        editor.putString(c.getResources().getString(R.string.preference_savedFont), getFontThemeID());

        if(localApply) {
            editor.apply();
        }
    }

    /**
     *
     * @param c
     * @param editor
     * @param localApply
     */
    private void saveAppTimeAlive(
            @NonNull Context c, @Nullable SharedPreferences.Editor editor, boolean localApply) {
        if(editor == null && (editor = getEditor(c)) == null) { return; }

        editor.putLong(c.getResources().getString(R.string.reviewtracking_appTimeAlive),
                getReviewRequestData().getTimeActive());

        if(localApply) {
            editor.apply();
        }
    }

    /**
     *
     * @param c
     * @param editor
     * @param localApply
     */
    private void saveTimesOpened(
            @NonNull Context c, @Nullable SharedPreferences.Editor editor, boolean localApply) {
        if(editor == null && (editor = getEditor(c)) == null) { return; }

        editor.putInt(c.getResources().getString(R.string.reviewtracking_appTimesOpened),
                getReviewRequestData().getTimesOpened());

        if(localApply) {
            editor.apply();
        }
    }

    /**
     *
     * @param c
     * @param editor
     * @param localApply
     */
    private void saveCanRequestReview(
            @NonNull Context c, @Nullable SharedPreferences.Editor editor, boolean localApply) {
        if(editor == null && (editor = getEditor(c)) == null) { return; }

        editor.putBoolean(c.getResources().getString(R.string.reviewtracking_canRequestReview),
                getReviewRequestData().getWasRequested());

        if(localApply) {
            editor.apply();
        }
    }

    /**
     *
     * @param c
     * @param editor
     * @param localApply
     */
    private void saveIsLeftHandSupportEnabled(
            @NonNull Context c, @Nullable SharedPreferences.Editor editor, boolean localApply) {
        if(editor == null && (editor = getEditor(c)) == null) { return; }

        editor.putBoolean(c.getResources().getString(R.string.preference_isLeftHandSupportEnabled),
                getIsLeftHandSupportEnabled());

        if(localApply) {
            editor.apply();
        }
    }

    /**
     *
     * @param c
     * @param editor
     * @param localApply
     */
    public void saveCanShowIntroduction(
            @NonNull Context c, @Nullable SharedPreferences.Editor editor, boolean localApply) {
        if(editor == null && (editor = getEditor(c)) == null) { return; }

        editor.putBoolean(c.getResources().getString(R.string.tutorialTracking_canShowIntroduction),
                getCanShowIntroduction());

        if(localApply) {
            editor.apply();
        }
    }

    /**
     * saveToFile method
     *
     * @param context The Activity context.
     */
    public void saveToFile(@NonNull Context context) {

        SharedPreferences.Editor editor = getEditor(context);

        saveNetworkPreference(context, editor, false);
        saveChosenLanguage(context, editor, false);
        saveAlwaysOnState(context, editor, false);
        saveHuntWarningAudioAllowed(context, editor, false);
        saveHuntWarningFlashTimeout(context, editor, false);
        saveColorSpace(context, editor, false);
        saveFontType(context, editor, false);
        saveIsLeftHandSupportEnabled(context, editor, false);
        saveCanRequestReview(context, editor, false);
        saveTimesOpened(context, editor, false);
        saveAppTimeAlive(context, editor, false);
        saveCanShowIntroduction(context, editor, false);

        editor.apply();
    }

    @NonNull
    public HashMap<String, String> getDataAsList() {
        HashMap<String, String> settings = new HashMap<>();
        settings.put("network_pref", String.valueOf(getNetworkPreference()));
        settings.put("language", getLanguageName());
        settings.put("always_on", String.valueOf(getIsAlwaysOn()));
        settings.put("warning_enabled", String.valueOf(getIsHuntAudioAllowed()));
        settings.put("warning_timeout", String.valueOf(getHuntWarningFlashTimeout()));
        settings.put("color_theme", getColorThemeID());
        settings.put("font_type", getFontThemeID());
        settings.put("left_support", String.valueOf(getIsLeftHandSupportEnabled()));
        settings.put("can_show_intro", String.valueOf(getCanShowIntroduction()));
        if(getReviewRequestData() != null) {
            settings.put("review_request", String.valueOf(getReviewRequestData().canRequestReview()));
            settings.put("times_opened", String.valueOf(getReviewRequestData().getTimesOpened()));
            settings.put("active_time", String.valueOf(getReviewRequestData().getTimeActive()));
        }

        return settings;
    }

    /**
     *
     * @param context
     */
    public void printFromFile(@NonNull Context context) {
        SharedPreferences sharedPref = getSharedPreferences(context);

        Log.d(context.getString(fileName),
                "NetworkPreference: " + sharedPref.getBoolean(context.getResources().getString(R.string.preference_network), getNetworkPreference()) +
                "; Language: " + sharedPref.getString(context.getResources().getString(R.string.preference_language), getLanguageName()) +
                "; Always On: " + sharedPref.getBoolean(context.getResources().getString(R.string.preference_isAlwaysOn), getIsAlwaysOn()) +
                "; Is Hunt Audio Allowed: " + sharedPref.getBoolean(context.getResources().getString(R.string.preference_isHuntAudioWarningAllowed), getIsHuntAudioAllowed()) +
                "; Hunt Warning Flash Timeout: " + sharedPref.getInt(context.getResources().getString(R.string.preference_huntWarningFlashTimeout), getHuntWarningFlashTimeout()) +
                "; Color Space: " + sharedPref.getString(context.getResources().getString(R.string.preference_savedTheme), getColorThemeID()) +
                "; ReviewRequestData: [" +
                "Time Alive: " + sharedPref.getLong(context.getResources().getString(R.string.reviewtracking_appTimeAlive), 0) +
                "; Times Opened: " + sharedPref.getInt(context.getResources().getString(R.string.reviewtracking_appTimesOpened), 0) +
                "; Can Request Review: " + sharedPref.getBoolean(context.getResources().getString(R.string.reviewtracking_canRequestReview), false) + "]" +
                "; Can Show Introduction: " + sharedPref.getBoolean(context.getResources().getString(R.string.tutorialTracking_canShowIntroduction), false));
    }

    /**
     *
     */
    public void printFromVariables() {
        Log.d("GlobalPreferencesVars",
                "NetworkPreference: " + getNetworkPreference() +
                "; Language: " + getLanguageName() +
                "; Always On: " + getIsAlwaysOn() +
                "; Is Hunt Audio Allowed: " + getIsHuntAudioAllowed() +
                "; Hunt Warning Flash Timeout: " + getHuntWarningFlashTimeout() +
                "; Color Space: " + getColorThemeID() +
                "; Can Show Introduction: " + getCanShowIntroduction() +
                "; ReviewRequestData: [" + getReviewRequestData().toString() + "]");
    }

}

File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/data/viewmodels/shared/GlobalPreferencesViewModel.kt
package com.TritiumGaming.phasmophobiaevidencepicker.data.viewmodels.shared

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import androidx.annotation.NonNull
import com.TritiumGaming.phasmophobiaevidencepicker.R
import com.TritiumGaming.phasmophobiaevidencepicker.data.controllers.ReviewTrackingData
import com.TritiumGaming.phasmophobiaevidencepicker.data.controllers.theming.CustomTheme
import com.TritiumGaming.phasmophobiaevidencepicker.data.controllers.theming.subsets.ColorThemeControl
import com.TritiumGaming.phasmophobiaevidencepicker.data.controllers.theming.subsets.FontThemeControl
import com.google.android.play.integrity.internal.c
import org.jetbrains.annotations.NotNull
import java.util.Locale

class GlobalPreferencesViewModel : SharedViewModel() {

    // Review Tracker
    lateinit var reviewRequestData: ReviewTrackingData
        private set

    // Persistent Styles
    lateinit var fontThemeControl: FontThemeControl
        private set
    lateinit var colorThemeControl: ColorThemeControl
        private set

    // Language
    var languageName: String = Locale.getDefault().language

    // Generic settings
    var huntWarningFlashTimeout: Int = -1 // Investigation behavior

    var isAlwaysOn: Boolean = false
    var isHuntWarningAudioAllowed: Boolean = true // Investigation behavior

    var networkPreference: Boolean = true
    var isLeftHandSupportEnabled: Boolean = false // Investigation behavior
    var reorderGhostViews = true // Investigation behavior

    // Title screen increments
    var canShowIntroduction: Boolean = true

    val colorThemeID: String
        get() = colorThemeControl.id
    val colorTheme: CustomTheme
        get() = colorThemeControl.currentTheme

    val fontThemeID: String
        get() = fontThemeControl.id
    val fontTheme: CustomTheme
        get() = fontThemeControl.currentTheme

    override fun setFileName() {
        fileName = R.string.preferences_globalFile_name
    }

    override fun init(context: Context): Boolean {
        setFileName()

        val sharedPref = getSharedPreferences(context)

        networkPreference =
            sharedPref.getBoolean(context.resources.getString(R.string.preference_network), networkPreference)
        languageName =
            sharedPref.getString(
                context.resources.getString(R.string.preference_language), languageName
            ) ?: Locale.getDefault().language
        isAlwaysOn =
            sharedPref.getBoolean(context.resources.getString(R.string.preference_isAlwaysOn), isAlwaysOn)
        isHuntWarningAudioAllowed =
            sharedPref.getBoolean(context.resources.getString(R.string.preference_isHuntAudioWarningAllowed), isHuntWarningAudioAllowed)
        huntWarningFlashTimeout =
            sharedPref.getInt(context.resources.getString(R.string.preference_huntWarningFlashTimeout), huntWarningFlashTimeout)

        isLeftHandSupportEnabled =
            sharedPref.getBoolean(context.resources.getString(R.string.preference_isLeftHandSupportEnabled), isLeftHandSupportEnabled)
        canShowIntroduction =
            sharedPref.getBoolean(context.resources.getString(R.string.tutorialTracking_canShowIntroduction), canShowIntroduction)

        reorderGhostViews =
            sharedPref.getBoolean(context.resources.getString(R.string.preference_enableReorderGhostViews), reorderGhostViews)


        reviewRequestData = ReviewTrackingData(
            sharedPref.getBoolean(context.resources.getString(R.string.reviewtracking_canRequestReview), false),
            sharedPref.getLong(context.resources.getString(R.string.reviewtracking_appTimeAlive), 0),
            sharedPref.getInt(context.resources.getString(R.string.reviewtracking_appTimesOpened), 0)
        )

        fontThemeControl = FontThemeControl(context)
        fontThemeControl.init(
            sharedPref.getString(context.resources.getString(R.string.preference_savedFont), fontThemeID)
        )

        colorThemeControl = ColorThemeControl(context)
        colorThemeControl.init(
            sharedPref.getString(context.resources.getString(R.string.preference_savedTheme), colorThemeID)
        )

        saveToFile(context)

        return true
    }

    fun incrementAppOpenCount(context: Context) {
        reviewRequestData.incrementTimesOpened()

        try {
            saveTimesOpened(context, getEditor(context), true)
        } catch (e: NullPointerException) {
            e.printStackTrace()
        }
    }


    /**
     * Gets the language saved to GlobalPreferences.
     * Defaults to return 'en' if there is no previously saved preference.
     *
     * @return The language specified in the Preferences data, or otherwise English
     */
    fun getLanguage(context: Context): String {
        val lang = context.getSharedPreferences(
            context.resources.getString(fileName), Context.MODE_PRIVATE
        ).getString("chosenLanguage", "en")

        Log.d("Current Chosen Language", lang!!)

        return lang
    }

    fun getLanguageIndex(languageNames: ArrayList<String?>): Int {
        for (i in languageNames.indices) {
            if (languageName.equals(languageNames[i], ignoreCase = true)) {
                return i
            }
        }
        return 0
    }

    fun setLanguage(position: Int, languageNames: Array<String?>) {
        if (position < 0 || position >= languageNames.size) {
            return
        }

        languageName = languageNames[position] ?: Locale.getDefault().language
    }

    fun canShowIntroduction(): Boolean {
        return canShowIntroduction && reviewRequestData!!.timesOpened <= 1
    }

    private fun saveNetworkPreference(
        context: Context, editor: SharedPreferences.Editor = getEditor(context), localApply: Boolean = false
    ) {
        editor.putBoolean(
            context.resources.getString(R.string.preference_network),
            networkPreference
        )

        if (localApply) {
            editor.apply()
        }
    }

    private fun saveChosenLanguage(
        context: Context, editor: SharedPreferences.Editor = getEditor(context), localApply: Boolean = false
    ) {
        editor.putString(
            context.resources.getString(R.string.preference_language),
            languageName
        )

        if (localApply) {
            editor.apply()
        }
    }

    private fun saveAlwaysOnState(
        context: Context, editor: SharedPreferences.Editor = getEditor(context), localApply: Boolean = false
    ) {
        editor.putBoolean(
            context.resources.getString(R.string.preference_isAlwaysOn),
            isAlwaysOn
        )

        if (localApply) {
            editor.apply()
        }
    }

    private fun saveHuntWarningAudioAllowed(
        context: Context, editor: SharedPreferences.Editor = getEditor(context), localApply: Boolean = false
    ) {
        editor.putBoolean(
            context.resources.getString(R.string.preference_isHuntAudioWarningAllowed),
            isHuntWarningAudioAllowed
        )

        if (localApply) {
            editor.apply()
        }
    }

    private fun saveHuntWarningFlashTimeout(
        context: Context, editor: SharedPreferences.Editor = getEditor(context), localApply: Boolean = false
    ) {
        editor.putInt(
            context.resources.getString(R.string.preference_huntWarningFlashTimeout),
            huntWarningFlashTimeout
        )

        if (localApply) {
            editor.apply()
        }
    }

    fun saveColorSpace(c: Context) {
        saveColorSpace(c, getEditor(c), true)
    }

    private fun saveColorSpace(
        context: Context, editor: SharedPreferences.Editor = getEditor(context), localApply: Boolean = false
    ) {
        editor.putString(
            context.resources.getString(R.string.preference_savedTheme),
            colorThemeID
        )

        if (localApply) {
            editor.apply()
        }
    }

    private fun saveFontType(
        context: Context, editor: SharedPreferences.Editor = getEditor(context), localApply: Boolean = false
    ) {
        editor.putString(context.resources.getString(R.string.preference_savedFont), fontThemeID)

        if (localApply) {
            editor.apply()
        }
    }

    private fun saveAppTimeAlive(
        context: Context, editor: SharedPreferences.Editor = getEditor(context), localApply: Boolean = false
    ) {
        editor.putLong(
            context.resources.getString(R.string.reviewtracking_appTimeAlive),
            reviewRequestData.timeActive
        )

        if (localApply) {
            editor.apply()
        }
    }

    private fun saveTimesOpened(
        context: Context, editor: SharedPreferences.Editor = getEditor(context), localApply: Boolean = false
    ) {
        editor.putInt(
            context.resources.getString(R.string.reviewtracking_appTimesOpened),
            reviewRequestData.timesOpened
        )

        if (localApply) {
            editor.apply()
        }
    }

    private fun saveCanRequestReview(
        context: Context, editor: SharedPreferences.Editor = getEditor(context), localApply: Boolean = false
    ) {
        editor.putBoolean(
            context.resources.getString(R.string.reviewtracking_canRequestReview),
            reviewRequestData.wasRequested
        )

        if (localApply) {
            editor.apply()
        }
    }

    private fun saveIsLeftHandSupportEnabled(
        context: Context, editor: SharedPreferences.Editor = getEditor(context), localApply: Boolean = false
    ) {
        editor.putBoolean(
            context.resources.getString(R.string.preference_isLeftHandSupportEnabled),
            isLeftHandSupportEnabled
        )

        if (localApply) {
            editor.apply()
        }
    }

    private fun saveReorderGhostViews(
        context: Context, editor: SharedPreferences.Editor = getEditor(context), localApply: Boolean = false
    ) {
        editor.putBoolean(
            context.resources.getString(R.string.preference_enableReorderGhostViews),
            reorderGhostViews
        )

        if (localApply) {
            editor.apply()
        }
    }

    fun saveCanShowIntroduction(
        context: Context, editor: SharedPreferences.Editor = getEditor(context), localApply: Boolean = false
    ) {
        editor.putBoolean(
            context.resources.getString(R.string.tutorialTracking_canShowIntroduction),
            canShowIntroduction
        )

        if (localApply) {
            editor.apply()
        }
    }

    override fun saveToFile(context: Context) {
        val editor = getEditor(context)

        saveNetworkPreference(context, editor, false)
        saveChosenLanguage(context, editor, false)
        saveAlwaysOnState(context, editor, false)
        saveHuntWarningAudioAllowed(context, editor, false)
        saveHuntWarningFlashTimeout(context, editor, false)
        saveColorSpace(context, editor, false)
        saveFontType(context, editor, false)
        saveIsLeftHandSupportEnabled(context, editor, false)
        saveReorderGhostViews(context, editor, false)
        saveCanRequestReview(context, editor, false)
        saveTimesOpened(context, editor, false)
        saveAppTimeAlive(context, editor, false)
        saveCanShowIntroduction(context, editor, false)

        editor.apply()
    }

    val dataAsList: HashMap<String, String?>
        get() {
            val settings = HashMap<String, String?>()
            settings["network_pref"] = networkPreference.toString()
            settings["language"] = languageName
            settings["always_on"] = isAlwaysOn.toString()
            settings["warning_enabled"] = isHuntWarningAudioAllowed.toString()
            settings["warning_timeout"] = huntWarningFlashTimeout.toString()
            settings["color_theme"] = colorThemeID
            settings["font_type"] = fontThemeID
            settings["left_support"] = isLeftHandSupportEnabled.toString()
            settings["can_show_intro"] = canShowIntroduction.toString()
            if (reviewRequestData != null) {
                settings["review_request"] = reviewRequestData!!.canRequestReview().toString()
                settings["times_opened"] = reviewRequestData!!.timesOpened.toString()
                settings["active_time"] = reviewRequestData!!.timeActive.toString()
            }

            return settings
        }

    fun printFromFile(context: Context) {
        val sharedPref = getSharedPreferences(context)

        Log.d(
            context.getString(fileName),
            "NetworkPreference: " + sharedPref.getBoolean(
                context.resources.getString(R.string.preference_network),
                networkPreference
            ) +
                    "; Language: " + sharedPref.getString(
                context.resources.getString(R.string.preference_language),
                languageName
            ) +
                    "; Always On: " + sharedPref.getBoolean(
                context.resources.getString(R.string.preference_isAlwaysOn),
                isAlwaysOn
            ) +
                    "; Is Hunt Audio Allowed: " + sharedPref.getBoolean(
                context.resources.getString(
                    R.string.preference_isHuntAudioWarningAllowed
                ), isHuntWarningAudioAllowed
            ) +
                    "; Hunt Warning Flash Timeout: " + sharedPref.getInt(
                context.resources.getString(
                    R.string.preference_huntWarningFlashTimeout
                ), huntWarningFlashTimeout
            ) +
                    "; Color Space: " + sharedPref.getString(
                context.resources.getString(R.string.preference_savedTheme),
                colorThemeID
            ) +
                    "; ReviewRequestData: [" +
                    "Time Alive: " + sharedPref.getLong(
                context.resources.getString(R.string.reviewtracking_appTimeAlive),
                0
            ) +
                    "; Times Opened: " + sharedPref.getInt(
                context.resources.getString(R.string.reviewtracking_appTimesOpened),
                0
            ) +
                    "; Can Request Review: " + sharedPref.getBoolean(
                context.resources.getString(R.string.reviewtracking_canRequestReview),
                false
            ) + "]" +
                    "; Can Show Introduction: " + sharedPref.getBoolean(
                context.resources.getString(
                    R.string.tutorialTracking_canShowIntroduction
                ), false
            )
        )
    }

    /**
     *
     */
    fun printFromVariables() {
        Log.d(
            "GlobalPreferencesVars",
            "NetworkPreference: " + networkPreference +
                    "; Language: " + languageName +
                    "; Always On: " + isAlwaysOn +
                    "; Is Hunt Audio Allowed: " + isHuntWarningAudioAllowed +
                    "; Hunt Warning Flash Timeout: " + huntWarningFlashTimeout +
                    "; Color Space: " + colorThemeID +
                    "; Can Show Introduction: " + canShowIntroduction +
                    "; ReviewRequestData: [" + reviewRequestData.toString() + "]"
        )
    }
}
